packages/superset-ui-preset-chart-xy/src/utils/XYChartLayout.tsx

```diff
@@ -0,0 +1,185 @@
+  chartHeight: number;
+  containerWidth: number;
+  containerHeight: number;
+  margin: Margin;
+  spec: Input;
+
+  xLayout?: {
+    labelOffset: number;
+    labellingStrategy: string;
+    rotation?: number;
+    tickTextAnchor?: string;
+    minMargin: Partial<Margin>;
+    orientation: string;
+  };
+
+  yLayout?: {
+    labelOffset: number;
+    minMargin: Partial<Margin>;
+    orientation: string;
+  };
+
+  // eslint-disable-next-line complexity
+  constructor(spec: Input) {
+    this.spec = spec;
+
+    const {
+      width,
+      height,
+      minContentWidth = 0,
+      minContentHeight = 0,
+      margin,
+      xEncoder,
+      yEncoder,
+      children,
+      theme,
+    } = spec;
+
+    const { xScale, yScale } = collectScalesFromProps({
+      width,
+      height,
+      margin,
+      xScale: xEncoder.definition.scale || {},
+      yScale: yEncoder.definition.scale || {},
+      theme,
+      children,
+    });
+
+    if (typeof yEncoder.axis !== 'undefined') {
+      const { axis: yAxis } = yEncoder;
+      this.yLayout = computeYAxisLayout({
+        orientation: yAxis.config.orient,
+        tickLabels: yAxis.getTickLabels(yScale),
+        tickLength: theme.yTickStyles.length,
+        tickTextStyle: theme.yTickStyles.label.right,
+      });
+    }
+
+    const secondMargin = this.yLayout ? mergeMargin(margin, this.yLayout.minMargin) : margin;
+    const innerWidth = Math.max(width - secondMargin.left - secondMargin.right, minContentWidth);
+
+    if (typeof xEncoder.axis !== 'undefined') {
+      const { axis: xAxis } = xEncoder;
+      const config = xAxis.config as XAxisConfig;
+      this.xLayout = computeXAxisLayout({
+        axisWidth: innerWidth,
+        orientation: config.orient || 'bottom',
+        labelAngle: config.labelAngle || this.recommendLabelAngle(config.orient),
+        tickLabels: xAxis.getTickLabels(xScale),
+        tickLength: theme.xTickStyles.length,
+        tickTextStyle: theme.xTickStyles.label.bottom,
+      });
+    }
+
+    const finalMargin = this.xLayout
+      ? mergeMargin(secondMargin, this.xLayout.minMargin)
+      : secondMargin;
+    const innerHeight = Math.max(height - finalMargin.top - finalMargin.bottom, minContentHeight);
+
+    const chartWidth = Math.round(innerWidth + finalMargin.left + finalMargin.right);
+    const chartHeight = Math.round(innerHeight + finalMargin.top + finalMargin.bottom);
+
+    const isOverFlowX = chartWidth > width;
+    const isOverFlowY = chartHeight > height;
+    if (isOverFlowX) {
+      finalMargin.bottom += OVERFLOW_MARGIN;
+    }
+    if (isOverFlowY) {
+      finalMargin.right += OVERFLOW_MARGIN;
+    }
+    this.chartWidth = isOverFlowX ? chartWidth + OVERFLOW_MARGIN : chartWidth;
+    this.chartHeight = isOverFlowY ? chartHeight + OVERFLOW_MARGIN : chartHeight;
+    this.containerWidth = width;
+    this.containerHeight = height;
+    this.margin = finalMargin;
+  }
+
+  recommendLabelAngle(xOrientation: 'top' | 'bottom' = 'bottom') {
```

Comment:
naming this `recommendXLabelAngle` or `getDefaultXLabelAngle` could clarify that it is only used for `XAxis`.